source/implementations/f7/Meadow.F7/Devices/F7CellNetworkAdapter.cs

```diff
@@ -206,5 +216,71 @@ public CellNetwork[] Scan()
     {
         return Core.Interop.Nuttx.MeadowCellNetworkScanner();
     }
+    
+    /// <summary>
+    /// Set the cell state
+    /// </summary>
+    /// <param name="CellState">State of the cell.</param>
+    private static void CellSetState (CellNetworkState CellState)
+    {
+        int state = 0;
+        switch (CellState)
+        {
+            case CellNetworkState.Resumed:
+                state = 0;
+                break;
+            case CellNetworkState.Paused:
+                state |= 1 << 0;
+                break;
+            case CellNetworkState.TrackingGPSLocation:
+                state |= 1 << 1;
+                break;
+            case CellNetworkState.FetchingSignalQuality:
+                state |= 1 << 2;
+                break;    
+            case CellNetworkState.ScanningNetworks:
+                state |= 1 << 3;
+                break;
+            default:
+                state |= 1 << 0;
+                break;
+        }
+        Core.Interop.Nuttx.meadow_cell_change_state(state);
+    }
 
+    /// <summary>
+    /// Return the the cell signal quality<b>CSQ<\b> if _at_cmds_ouput is null, otherwise <b>99</b>
+    /// </summary>
+    public double GetSignalQuality()
+    {
+        string csqPattern = @"\+CSQ:\s+(\d+),\d+";
+        string csq;
+        _at_timeout = 3;
+
+        CellSetState(CellNetworkState.FetchingSignalQuality);
+
+        while (_at_timeout > 0 || !_at_cmd_done)
```

Comment:
This while seems to be not necessary since the timeout will be handled by the OS, based on the PPP timeout.